### PR TITLE
Add nightly cron jobs for openAPI branch

### PR DESCRIPTION
**Pull request by DunnCoding**
This PR adds the nightly chron builds to the `openAPI` branch as well as the `init` branch. 

I've attempted to do this in a way to future proof the addition of other branches in the future. 

As a part of this PR I also removed any instances of using `cd ../` and used absolute paths instead.

### CODE REVIEW

**Review by ianpartridge — COMMENTED** [review]
Before we do a detailed review, please run this through shellcheck.net and fix the issues it finds.

**Comment by DunnCoding** [comment]
@ianpartridge I entered the script into `spellcheck.net` and have resolved any warnings it revealed.

File: updateRepo.sh [hunk 1]
```diff
@@ -1,34 +1,53 @@
+#!/bin/bash
+
 echo "Checking if repo needs to be updated"
 export ORG="IBM-Swift"
 export REPO="generator-swiftserver-projects"
 export GH_REPO="github.com/${ORG}/${REPO}.git"
-export BRANCH="init"
+export BRANCHES="init openAPI"
```
Comment (ianpartridge) [inline updateRepo.sh]: Why do all these need to be exported?

File: updateRepo.sh [hunk 2]
```diff
@@ -1,34 +1,53 @@
...
-export currentProject=`pwd`
+for BRANCH in $BRANCHES
+do
+  cd "${TRAVIS_BUILD_DIR}" || exit
+  mkdir current
+  cd current || exit
+  git clone -b "${BRANCH}" "https://${GITHUB_USERNAME}:${GITHUB_PASSWORD}@github.com/${ORG}/${REPO}.git"
+  currentProject=$(pwd)
+
+  mkdir -p "${TRAVIS_BUILD_DIR}/new/${projectName}"
```
Comment (ianpartridge) [inline updateRepo.sh]: Why are we creating TRAVIS_BUILD_DIR, aren't we already in that directory?

File: updateRepo.sh [hunk 3]
```diff
@@ -1,34 +1,53 @@
...
+  cd "${TRAVIS_BUILD_DIR}/new/${projectName}" || exit
+  if [[ "${BRANCH}" == "init" ]]
+  then
+    yo swiftserver --init --skip-build
+  elif [[ "$BRANCH" == "openAPI" ]]
+  then
+    yo swiftserver --app --spec '{ "appName":"'"$projectName"'", "appType":"scaffold", "appDir":".", "openapi":true, "docker":true, "metrics":true, "healthcheck":true }' --skip-build
+  fi
```
Comment (ianpartridge) [inline updateRepo.sh:49]: Can we have a final `else` that ends the script?

File: updateRepo.sh [hunk 4]
```diff
@@ -1,34 +1,53 @@
...
+
+  rm spec.json
```
Comment (ianpartridge) [inline updateRepo.sh]: Do we need to do this in the BRANCH=init case?

File: updateRepo.sh [hunk 5]
```diff
@@ -1,34 +1,53 @@
...
-cd ${currentProject}/${REPO}
-git add .
-git commit -m "CRON JOB: Updating generated project"
-git push origin ${BRANCH}
+  echo "Project needs to be updated"
+  cp -r "${newProject}/${projectName}/." "${currentProject}/${REPO}"
+  cd "${currentProject}/${REPO}" || exit
+  git add .
+  git commit -m "CRON JOB: Updating generated project"
+  git push origin "${BRANCH}"
```
Comment (ianpartridge) [inline updateRepo.sh]: Need to handle failure here.

File: updateRepo.sh [hunk 6]
```diff
@@ -1,34 +1,53 @@
...
-echo "Project needs to be updated"
-cp -r ${newProject}/${projectName}/. ${currentProject}/${REPO}
-cd ${currentProject}/${REPO}
-git add .
-git commit -m "CRON JOB: Updating generated project"
-git push origin ${BRANCH}
+  echo "Project needs to be updated"
+  cp -r "${newProject}/${projectName}/." "${currentProject}/${REPO}"
+  cd "${currentProject}/${REPO}" || exit
+  git add .
```
Comment (ianpartridge) [inline updateRepo.sh]: Would this handle the case where a file is deleted from the project?

File: updateRepo.sh (hunk 2 above)
Comment (DunnCoding) [reply updateRepo.sh]: This relates to work prior to my PR so I'm not 100% sure, previously it used `../` I felt this was a better approach. 
At this stage of the script though we are in the following directory: `${TRAVIS_BUILD_DIR}/current`

File: updateRepo.sh (hunk 1 above)
Comment (DunnCoding) [reply updateRepo.sh]: I don't think they do, I'm not 100% sure why they were exported in the first place but I can't see that they're being used elsewhere.

File: updateRepo.sh (hunk 4 above)
Comment (DunnCoding) [reply updateRepo.sh]: Good spot, we do not need to do this in that case. I'll fix now.

File: updateRepo.sh (hunk 2 above)
Comment (ianpartridge) [reply updateRepo.sh]: I'm confused.  On line 12 we `cd` into `${TRAVIS_BUILD_DIR}` but on line 18 we create it?

File: updateRepo.sh (hunk 6 above)
Comment (DunnCoding) [reply updateRepo.sh]: I don't think so, nor would it handle a new file I think. 
Again I wasn't why this was done that way before but I'll correct this to account for those cases.

File: updateRepo.sh (hunk 1 above)
Comment (ianpartridge) [reply updateRepo.sh]: OK cool, let's fix this script up properly then :)

File: updateRepo.sh (hunk 3 above)
Comment (DunnCoding) [reply updateRepo.sh:49]: Will add that, thanks!

File: updateRepo.sh (hunk 2 above)
Comment (DunnCoding) [reply updateRepo.sh]: Ah I understand now. 

The `TRAVIS_BUILD_DIR` isn't something we've created, it's an environment variable for Travis. We just use it to ensure we're creating sub directories in a known location.

Comment (ianpartridge) [reply updateRepo.sh]: But this line will create a subdirectory tree with that name, won't it?

Comment (ianpartridge) [reply updateRepo.sh]: `mkdir` works relative to the current directory, it's not absolute.

Comment (ianpartridge) [reply updateRepo.sh]: Or is `${TRAVIS_BUILD_DIR}` absolute? If so, we don't need to specify it here do we?

Comment (DunnCoding) [reply updateRepo.sh]: I'm not entirely sure what you mean here. 

The command in question: 
`mkdir -p "${TRAVIS_BUILD_DIR}/new/${projectName}"`

Creates the `new/${projectName}"` directory within the `${TRAVIS_BUILD_DIR}` directory. Which is what the original script was intending to do, but the original script just used `../` instead. It does not recreate the `${TRAVIS_BUILD_DIR}` within itself. 

I've tested this in the [following build](https://travis-ci.org/ddunn2/generator-swiftserver-projects/builds/479281458#L1098) where after creating the directory I attempt to `cd` into the `new` directory to ensure it has been created in the correct location. **Note:** I have to step up one level as we are still inside the `current` directory.

**Review by djones6 — COMMENTED** [review]
Script looks good, but please can you confirm that your latest `rsync` command is doing what you expect (it appears so from the Travis output, but I haven't dug deep)

**Comment by DunnCoding** [comment]
@djones6 It seems to be doing what we'd expect. 

I think this is good to merge.

File: updateRepo.sh [hunk 7]
```diff
@@ -1,34 +1,68 @@
...
+  newProject="$(pwd)/new"
+
+  currentRepo="${currentProject}/${REPO}"
+  newRepo="${newProject}/${projectName}"
+
+  if diff -x '.git' -r "${currentRepo}" "${newRepo}"
+  then
+    echo "Project does not need to be updated"
+    rm -rf "${currentRepo}" "${newRepo}"
+    exit 1
```
Comment (djones6) [inline updateRepo.sh]: This doesn't seem right - shouldn't this be `continue`?
You may have other repos in the enclosing `for` loop to try.  Another symptom is that the final output from the script "Success: <list of branches" doesn't appear in your build log: https://travis-ci.org/IBM-Swift/generator-swiftserver-projects/builds/511976963

**Comment by djones6** [comment]
@ianpartridge FYI, the solution to preventing unwanted pushes to 'init' etc is:
1. Turn off Travis from auto-building PRs
2. Add a check to the script to avoid trying to push for a PR build (belt & braces - ie. in case the build was mistakenly re-enabled)

This means that to see the results of a script change, we'll require people to fork the repo and run the CI through their own Travis (having defined their credentials).

We can avoid people having to change the org name in the script by using `TRAVIS_REPO_SLUG` instead. @ddunn2 is going to make that change.

**Pull request merged**